File: sdk/cpp/src/heartbeat.cpp
#include "cap/heartbeat.h"
#include "cap/codec.h"
#include "cap/subjects.h"

#include <google/protobuf/util/time_util.h>

#include "cordum/agent/v1/buspacket.pb.h"
#include "cordum/agent/v1/heartbeat.pb.h"

namespace cap {

std::vector<uint8_t> HeartbeatPayload(
    const std::string& worker_id,
    const std::string& pool,
    int32_t active_jobs,
    int32_t max_parallel,
    float cpu_load) {
  return HeartbeatPayloadWithProgress(
      worker_id, pool, active_jobs, max_parallel, cpu_load, 0.0f, 0, "");
}

std::vector<uint8_t> HeartbeatPayloadWithMemory(
    const std::string& worker_id,
    const std::string& pool,
    int32_t active_jobs,
    int32_t max_parallel,
    float cpu_load,
    float memory_load) {
  return HeartbeatPayloadWithProgress(
      worker_id, pool, active_jobs, max_parallel, cpu_load, memory_load, 0, "");
}

std::vector<uint8_t> HeartbeatPayloadWithProgress(
    const std::string& worker_id,
    const std::string& pool,
    int32_t active_jobs,
    int32_t max_parallel,
    float cpu_load,
    float memory_load,
    int32_t progress_pct,
    const std::string& last_memo) {
  cordum::agent::v1::BusPacket packet;
  packet.set_sender_id(worker_id);
  packet.set_protocol_version(1);
  packet.mutable_created_at()->CopyFrom(
      google::protobuf::util::TimeUtil::GetCurrentTime());

  auto* hb = packet.mutable_heartbeat();
  hb->set_worker_id(worker_id);
  hb->set_pool(pool);
  hb->set_active_jobs(active_jobs);
  hb->set_max_parallel_jobs(max_parallel);
  hb->set_cpu_load(cpu_load);
  hb->set_memory_load(memory_load);
  hb->set_progress_pct(progress_pct);
  hb->set_last_memo(last_memo);

  return MarshalDeterministic(packet);
}

bool EmitHeartbeat(BusClient* bus, const std::vector<uint8_t>& payload) {
  if (!bus) return false;
  return bus->Publish(kSubjectHeartbeat, payload);
}

HeartbeatLoop::HeartbeatLoop(BusClient* bus,
                             PayloadFn payload_fn,
                             std::string worker_id,
                             std::chrono::milliseconds interval,
                             MetricsHook* metrics)
    : bus_(bus),
      payload_fn_(std::move(payload_fn)),
      worker_id_(std::move(worker_id)),
      interval_(interval),
      metrics_(metrics) {
  thread_ = std::thread(&HeartbeatLoop::Run, this);
}

HeartbeatLoop::~HeartbeatLoop() {
  Stop();
}

void HeartbeatLoop::Stop() {
  stop_.store(true, std::memory_order_release);
  if (thread_.joinable()) {
    thread_.join();
  }
}

void HeartbeatLoop::Run() {
  while (!stop_.load(std::memory_order_acquire)) {
    auto payload = payload_fn_();
    if (EmitHeartbeat(bus_, payload) && metrics_) {
      metrics_->OnHeartbeatSent(worker_id_);
    }

    // Sleep in small increments to respond quickly to Stop().
    auto deadline = std::chrono::steady_clock::now() + interval_;
    while (!stop_.load(std::memory_order_acquire) &&
           std::chrono::steady_clock::now() < deadline) {
      std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }
  }
}

}  // namespace cap

File: sdk/php/src/Heartbeat.php
<?php

declare(strict_types=1);

namespace Cordum\Cap;

use Basis\Nats\Client as NatsClient;
use Cordum\Agent\V1\BusPacket;
use Cordum\Agent\V1\Heartbeat as HeartbeatMessage;
use Google\Protobuf\Timestamp;

/**
 * Heartbeat payload constructors and emission helpers.
 */
final class HeartbeatHelper
{
    /**
     * Build a heartbeat payload with basic fields.
     */
    public static function heartbeatPayload(
        string $workerId,
        string $pool,
        int $activeJobs,
        int $maxParallel,
        float $cpuLoad,
    ): string {
        return self::heartbeatPayloadWithProgress($workerId, $pool, $activeJobs, $maxParallel, $cpuLoad, 0.0, 0, '');
    }

    /**
     * Build a heartbeat payload with memory load.
     */
    public static function heartbeatPayloadWithMemory(
        string $workerId,
        string $pool,
        int $activeJobs,
        int $maxParallel,
        float $cpuLoad,
        float $memoryLoad,
    ): string {
        return self::heartbeatPayloadWithProgress($workerId, $pool, $activeJobs, $maxParallel, $cpuLoad, $memoryLoad, 0, '');
    }

    /**
     * Build a heartbeat payload with all fields including progress and memo.
     */
    public static function heartbeatPayloadWithProgress(
        string $workerId,
        string $pool,
        int $activeJobs,
        int $maxParallel,
        float $cpuLoad,
        float $memoryLoad,
        int $progressPct,
        string $lastMemo,
    ): string {
        $hb = new HeartbeatMessage();
        $hb->setWorkerId($workerId);
        $hb->setPool($pool);
        $hb->setActiveJobs($activeJobs);
        $hb->setMaxParallelJobs($maxParallel);
        $hb->setCpuLoad($cpuLoad);
        $hb->setMemoryLoad($memoryLoad);
        $hb->setProgressPct($progressPct);
        $hb->setLastMemo($lastMemo);

        $now = microtime(true);
        $timestamp = new Timestamp();
        $timestamp->setSeconds((int) $now);
        $timestamp->setNanos((int) (($now - (int) $now) * 1_000_000_000));

        $packet = new BusPacket();
        $packet->setSenderId($workerId);
        $packet->setProtocolVersion(Subjects::DEFAULT_PROTOCOL_VERSION);
        $packet->setCreatedAt($timestamp);
        $packet->setHeartbeat($hb);

        return Codec::marshalDeterministic($packet);
    }

    /**
     * Publish a heartbeat payload to the heartbeat subject.
     */
    public static function emitHeartbeat(NatsClient $nc, string $payload): void
    {
        $nc->publish(Subjects::HEARTBEAT, $payload);
    }
}

/**
 * Periodically emits heartbeats using a timer loop.
 * Uses sleep-based loop since PHP is single-threaded.
 */
class HeartbeatLoop
{
    private bool $running = false;

    /**
     * @param NatsClient $nc NATS connection
     * @param callable $payloadFn Function returning heartbeat payload bytes
     * @param string $workerId Worker ID for metrics
     * @param int $intervalSeconds Heartbeat interval in seconds
     * @param MetricsHook $metrics Metrics hook
     */
    public function __construct(
        private readonly NatsClient $nc,
        private readonly mixed $payloadFn,
        private readonly string $workerId,
        private readonly int $intervalSeconds = Subjects::DEFAULT_HEARTBEAT_INTERVAL,
        private readonly MetricsHook $metrics = new NoopMetrics(),
    ) {}

    /**
     * Run the heartbeat loop. Blocks until stop() is called.
     * Best used with pcntl_signal for clean shutdown.
     */
    public function start(): void
    {
        $this->running = true;
        while ($this->running) {
            try {
                $payload = ($this->payloadFn)();
                HeartbeatHelper::emitHeartbeat($this->nc, $payload);
                $this->metrics->onHeartbeatSent($this->workerId);
            } catch (\Throwable) {
                // Swallow individual emit failures
            }
            sleep($this->intervalSeconds);
        }
    }

    /**
     * Emit a single heartbeat tick. Useful for integration with external event loops.
     */
    public function tick(): void
    {
        try {
            $payload = ($this->payloadFn)();
            HeartbeatHelper::emitHeartbeat($this->nc, $payload);
            $this->metrics->onHeartbeatSent($this->workerId);
        } catch (\Throwable) {
            // Swallow
        }
    }

    /**
     * Stop the heartbeat loop.
     */
    public function stop(): void
    {
        $this->running = false;
    }
}
